src/mscorlib/src/System/ParseNumbers.cs

```diff
@@ -588,21 +651,41 @@ private static int GrabInts(int radix, string s, ref int i, bool isUnsigned)
             return (int)result;
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Is this needed? I believe that the JIT is smart enough to not inline methods that just throw. And if you do not mark the throw helpers with NoInlining, it will actually generate slightly better code because of it will realize that this is cold path and that the call never returns.

Reply:
> Is this needed? I believe that the JIT is smart enough to not inline methods that just throw. 

Ah, I'll try. I separated it out and added the attribute at the same time; that helped, but never tried without the attribute.

>  it will actually generate slightly better code because of it will realize that this is cold path and that the call never returns

Why does the attribute inhibit that?

Reply:
The attribute makes the JIT to not look at the inlinee at all. Without the attribute, the JIT does look at the inlinee, tries to inline it and even though it does not end up inlining it at the end, it will use the extra evidence gathered during the inlining attempt for further optimizations.

cc @AndyAyersMS

Reply:
Thanks. Makes sense, although it's counter-intuitive to me. It seems like the JIT could still gather that same evidence either way, even if there's a small JIT throughput win to be had by skipping it.

Reply:
Jan has it right:
* with the noinline attribute in place, the jit immediately rejects the inline and doesn't look at the inlinee at all. So it doesn't realize the method must throw, and doesn't mark the path with the call as cold, etc.
* without the attribute the jit realizes the inlinee must throw and rejects the inline and marks the path leading to the call as "rarely taken". Usual policy in similar cases is for the jit to then mark the method descriptor as not inlinable to short-circut future inline attempts, but in this particular case the jit defers doing so. Instead at at each call it will re-examine the method and do the same reject/rarely-taken marking.

If we had a more evolved system for recording jit "interprocedural" observations we could possibly rework this and have the jit propagate "noinline noretrurn" back to the method descriptor, so that after the first examination of the method body the jit could both quickly reject the inline and mark the path as rarely taken. This sort of thing is part of #6095.